prow/pjutil/abort.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pjutil
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+// prowClient a minimalistic prow client required by the aborter
+type prowClient interface {
+	//ReplaceProwJob replaces the prow job with the given name
+	ReplaceProwJob(string, prowapi.ProwJob) (prowapi.ProwJob, error)
+}
+
+// ProwJobResourcesCleanup type for a callback function which it is expected to clean up
+// all k8s resources associated with the given prow job. It should do the best effort to
+// remove these resources, but if for any reason there is an error, it should only log a warning
+// message.
+type ProwJobResourcesCleanup func(pj prowapi.ProwJob)
+
+// ProwJobAborter is an interface for abstracting the prow job aborter behaviour
+type ProwJobAborter interface {
+	// TerminateOlderPresubmitJobs aborts all prow presubmit jobs from the given list that
+	// have a newer version, and call the callback on each aborted job
+	TerminateOlderPresubmitJobs(pjs []prowapi.ProwJob, cleanup ProwJobResourcesCleanup) error
+}
+
+// ProwJobAborter provides functionality to abort prow jobs
+type prowJobAborter struct {
+	pjc prowClient
+	log *logrus.Entry
+}
+
+// jobIndentifier keeps the information required to uniquely identify a prow job
+type jobIndentifier struct {
+	job          string
+	organization string
+	repository   string
+	pullRequest  int
+}
+
+// String returns the string representation of a prow job identifier
+func (i *jobIndentifier) String() string {
+	return fmt.Sprintf("%s %s/%s#%d", i.job, i.organization, i.repository, i.pullRequest)
+}
+
+//NewProwJobAborter creates a new ProwJobAborter
+func NewProwJobAborter(pjc prowClient, log *logrus.Entry) ProwJobAborter {
+	return &prowJobAborter{
+		log: log,
+		pjc: pjc,
+	}
+}
+
+// TerminateOlderPresubmitJobs aborts all presubmit jobs from the given list that have a newer version. It calls
+// the cleanup callback for each job before updating its status as aborted.
+func (a *prowJobAborter) TerminateOlderPresubmitJobs(pjs []prowapi.ProwJob, cleanup ProwJobResourcesCleanup) error {
```

Comment:
There is only a single `TerminateOlderPresubmitJobs` implementation. The only thing tthat changes is whether we delete pods or builds (hence the cleanup argument). All the rest of the abstraction here feels unnecessary.

A better design is to just make this a function rather than a method: `func TerminateOlderPresubmits(pjc prowClient, log *logrus.Entry, pjs []prowapi.ProwJob, cleanup ProwJobResourcesCleanup) error`.

I do not see any benefit from the unnecessary structures, interfaces, etc.

This is how plank and build share functionality like https://github.com/kubernetes/test-infra/blob/f58913b81191b5ff9e59a2edcb5691081e7d1819/prow/pod-utils/decorate/podspec.go#L280